libnopegl/src/node_hblur.c

```diff
@@ -0,0 +1,677 @@
+        .usage         = NGLI_TEXTURE_USAGE_COLOR_ATTACHMENT_BIT |
+                         NGLI_TEXTURE_USAGE_SAMPLED_BIT,
+    };
+
+    ret = ngli_texture_init(tex0, &texture_params);
+    if (ret < 0)
+        goto fail;
+
+    ret = ngli_texture_init(tex1, &texture_params);
+    if (ret < 0)
+        goto fail;
+
+    struct rtt_params pass1_rtt_params = {
+        .width = width,
+        .height = height,
+        .nb_colors = 2,
+        .colors[0] = {
+            .attachment = tex0,
+            .store_op = NGLI_STORE_OP_STORE
+        },
+        .colors[1] = {
+            .attachment = tex1,
+            .store_op = NGLI_STORE_OP_STORE
+        },
+    };
+
+    ret = ngli_rtt_init(pass1_rtt_ctx, &pass1_rtt_params);
+    if (ret < 0)
+        goto fail;
+
+    /* Assert that the destination texture format does not change */
+    struct texture_priv *dst_priv = o->destination->priv_data;
+    ngli_assert(dst_priv->params.format == s->pass2.layout.colors[0].format);
+
+    dst = dst_priv->texture;
+    if (s->dst_is_resizeable) {
+        dst = ngli_texture_create(ctx->gpu_ctx);
+        if (!dst) {
+            ret = NGL_ERROR_MEMORY;
+            goto fail;
+        }
+
+        struct texture_params params = dst_priv->params;
+        params.width = width;
+        params.height = height;
+        ret = ngli_texture_init(dst, &params);
+        if (ret < 0)
+            goto fail;
+    }
+
+    if (s->dst_is_resizeable) {
+        ngli_texture_freep(&dst_priv->texture);
+        dst_priv->texture = dst;
+        dst_priv->image.params.width = dst->params.width;
+        dst_priv->image.params.height = dst->params.height;
+        dst_priv->image.planes[0] = dst;
+        dst_priv->image.rev = dst_priv->image_rev++;
+    }
+
+    pass2_rtt_ctx = ngli_rtt_create(ctx);
+    if (!pass2_rtt_ctx) {
+        ret = NGL_ERROR_MEMORY;
+        goto fail;
+    }
+
+    const struct rtt_params pass2_rtt_params = (struct rtt_params) {
+        .width  = dst->params.width,
+        .height = dst->params.height,
+        .nb_colors = 1,
+        .colors[0] = {
+            .attachment = dst,
+            .load_op = NGLI_LOAD_OP_CLEAR,
+            .store_op = NGLI_STORE_OP_STORE,
+        },
+    };
+
+    ret = ngli_rtt_init(pass2_rtt_ctx, &pass2_rtt_params);
+    if (ret < 0)
+        goto fail;
+
+    ngli_rtt_freep(&s->pass2.rtt_ctx);
+    s->pass2.rtt_ctx = pass2_rtt_ctx;
+
+    ngli_rtt_freep(&s->pass1.rtt_ctx);
+    s->pass1.rtt_ctx = pass1_rtt_ctx;
+
+    ngli_texture_freep(&s->tex0);
+    s->tex0 = tex0;
+
+    ngli_texture_freep(&s->tex1);
+    s->tex1 = tex1;
+
+    ngli_pipeline_compat_update_texture(s->pass2.pl, 0, s->tex0);
+    ngli_pipeline_compat_update_texture(s->pass2.pl, 1, s->tex1);
+
+    s->width = width;
+    s->height = height;
+
+    return 0;
+
+fail:
+    ngli_rtt_freep(&pass1_rtt_ctx);
+    ngli_texture_freep(&tex0);
+    ngli_texture_freep(&tex1);
+
+    ngli_rtt_freep(&pass2_rtt_ctx);
```

Comment:
any reason to separate `pass[12]_rtt_ctx` like that?

Reply:
Yep, the goal is to pack the rtt_ctx with its render targets (the textures).